daft/utils.py

```diff
@@ -124,3 +125,38 @@ def column_inputs_to_expressions(columns: ManyColumnsInputType) -> list[Expressi
 
     column_iter: Iterable[ColumnInputType] = [columns] if is_column_input(columns) else columns  # type: ignore
     return [col(c) if isinstance(c, str) else c for c in column_iter]
+
+
+class SyncFromAsyncIterator:
+    """Convert an async iterator to a sync iterator.
+
+    Note that the async iterator is created lazily upon first iteration.
+    """
+
+    def __init__(self, async_iter_producer: Callable[[], AsyncIterator]):
+        self.async_iter_producer = async_iter_producer
+        self.async_iter = None
+        self.loop = asyncio.new_event_loop()
+        self.stopped = False
+
+    def __iter__(self):
+        return self
```

Comment:
raise an error if it's already stopped or if iter has been called on this before